### ISSUE TITLE
UNIQUE constraint failed

### ISSUE THREAD

**Issue by boramalper** [boramalper/magnetico#97]
I got the following error this morning:

```
2017-06-13 06:35:24,576  INFO      Added: `REDACTED`
2017-06-13 06:35:24,578  ERROR     Could NOT commit metadata to the database! (30122 metadata are pending)
Traceback (most recent call last):
  File "/home/bora/.local/lib/python3.5/site-packages/magneticod/persistence.py", line 116, in __commit_metadata
    self.__pending_metadata
sqlite3.IntegrityError: UNIQUE constraint failed: torrents.info_hash
2017-06-13 06:35:25,784  INFO      Added: `REDACTED`
2017-06-13 06:35:25,785  ERROR     Could NOT commit metadata to the database! (30123 metadata are pending)
Traceback (most recent call last):
  File "/home/bora/.local/lib/python3.5/site-packages/magneticod/persistence.py", line 116, in __commit_metadata
    self.__pending_metadata
sqlite3.IntegrityError: UNIQUE constraint failed: torrents.info_hash
2017-06-13 06:35:26,301  INFO      Peer does NOT support the extension protocol
2017-06-13 06:35:26,965  INFO      Peer does NOT support the extension protocol
^C2017-06-13 06:35:27,032  CRITICAL  Keyboard interrupt received! Exiting gracefully...
2017-06-13 06:35:27,033  ERROR     Could NOT commit metadata to the database! (30123 metadata are pending)
Traceback (most recent call last):
  File "/home/bora/.local/lib/python3.5/site-packages/magneticod/persistence.py", line 116, in __commit_metadata
    self.__pending_metadata
sqlite3.IntegrityError: UNIQUE constraint failed: torrents.info_hash
```

My initial guess is that there is a race condition that causes the issue: What if two child (fetch) tasks finish at the same time?

**Comment by ad-m** [comment]
How much are you convinced of this idea? Is our application uses multiple threads?

**Comment by boramalper** [comment]
I was thinking what would happen if two child tasks finish at the same time but that shouldn't cause the problem either. See https://github.com/boramalper/magnetico/blob/master/magneticod/magneticod/dht.py#L275

**Comment by ad-m** [comment]
Is our application parallel? Is there any [preemption](https://en.wikipedia.org/wiki/Preemption_(computing)) of tasks / threads?

**Comment by boramalper** [comment]
As far as I know, no, but @richardkiss might shed some light here. Asyncio is cooperative rather than preemptive, but that is also tricky enough in itself. The only guarantee you have is that a function/method will not be interrupted during its execution, and might be interrupted only while it's `await`ing, but still you don't know what might be scheduled next by the event loop. An instance of surprising behaviour is in the link I shared above, and I am suspicious of a similar case.

**Comment by richardkiss** [comment]
There is no pre-emption or threads (with the exception of the built-in subtask of asyncio which does DNS resolution).

**Comment by boramalper** [comment]
@Bentzia you wrote that you also experience this issue; is it happening regularly, or happened once or twice?

**Comment by Bentzia** [comment]
I realized the problem this morning and just restarted magneticod.
It seems working smoothly now. I'll let magneticod running until tomorrow and I'll tell you how it goes. 

These are the versions I'm using.
`$ uname -a`
`Linux 4.9.24-v7+ #993 SMP Wed Apr 26 18:01:23 BST 2017 armv7l GNU/Linux`
`$ ls -l $(which python)`
`lrwxrwxrwx 1 root staff 9 jun 13 14:55 /usr/local/bin/python -> python3.5`
`$ ls -l $(which python)3.5`
`-rwxr-xr-x 2 root staff 9243704 jun 14 09:15 /usr/local/bin/python3.5`
`$ python -V`
`Python 3.5.3`
`$ magneticod`
`2017-06-15 10:53:57,135  INFO      magneticod v0.6.0 started`

python installed from sources
On a Raspberry Pi 3 with raspbian Jessie

If you need any more information from me, just ask for it.

**Comment by Glandos** [comment]
I also got the issue this morning, and the CPU usage of magneticod went berserk. Restarting magneticod solved the issue.

**Comment by Bentzia** [comment]
After tow days after restarting the issue doesn't repeats.

**Comment by Glandos** [comment]
I've got the error 3 times in a row yesterday morning, and I've added more traces to the code. But now, I'm still waiting for this to happens…

**Comment by boramalper** [comment]
```
2017-06-17 06:32:29,508  ERROR     Could NOT commit metadata to the database! (62465 metadata are pending)
Traceback (most recent call last):
  File "/home/bora/.local/lib/python3.5/site-packages/magneticod/persistence.py", line 119, in __commit_metadata
    self.__pending_metadata
sqlite3.IntegrityError: CHECK constraint failed: torrents
```

I just woke up to see that I lost 60k torrents because of this error... I'll work on it this weekend.

**Comment by boramalper** [comment]
Okay, the revelation just came now:

* One of the child tasks fetch the metadata, kill all other children, and add the metadata to the queue.
* While the metadata stays the queue to be "got" by the `metadata_queue_watcher()`, an `announce_peer` query with the same info hash is received! Since there are no pending tasks on the info hash (as we killed them all) nor any record in the `class Database()` instance (as the metadata is still in the `metadata_queue: asyncio.Queue`), it is regarded as a new info hash and another task is created.
* Of course, when the first metadata is gotten from the queue and flushed to the database, the second metadata of the same info hash will arrive too and cause the IntegrityError.

It also explains how rare the bug is, since it depends on the timing and the unknown behaviour of the underlying event loop (how coroutines are scheduled). Besides, the second metadata should also be received from the remote peer successfully before the timeout, which makes it even rarer!

__It made me feel great that now I found it. =)__
